src/main/java/com/vi/counselingtoolsservice/api/service/budibase/BudibaseProxyService.java

```diff
@@ -0,0 +1,206 @@
+import org.springframework.http.HttpMethod;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Service;
+import org.springframework.web.client.RestTemplate;
+import org.springframework.web.util.UriComponentsBuilder;
+
+@Service
+@Slf4j
+@RequiredArgsConstructor
+public class BudibaseProxyService {
+
+  @Value("${budibase.proxy.host}")
+  private String proxyServiceHost;
+
+  @Value("${budibase.proxy.port}")
+  private Integer proxyServicePort;
+
+  @Value("${budibase.proxy.whitelisted}")
+  private String whitelistedURIs;
+
+  private final BudibaseApiService budibaseApiService;
+
+  public boolean isWhiteListed(HttpServletRequest request) {
+    Set<String> whitelistedURIs = new HashSet<>(Arrays.asList(this.whitelistedURIs.split(";")));
+    Optional<String> match = whitelistedURIs.stream()
+        .filter(el -> request.getRequestURI().contains(el)).findFirst();
+    return match.isPresent();
+  }
+
+  public String extractUserIdFromJWT(HttpServletRequest request) {
+    Base64.Decoder decoder = Base64.getUrlDecoder();
+
+    String[] cookies = request.getHeader("cookie").split(";");
+
+    Optional<String> authCookie = Arrays.stream(cookies)
+        .filter(cookie -> cookie.contains("budibase:auth") && !cookie.equals("budibase:auth="))
+        .findFirst();
+
+    String token = authCookie.get().split("=")[1];
+    String[] chunks = token.split("\\.");
+    String payload = new String(decoder.decode(chunks[1]));
+    JSONObject jsonObject = new JSONObject(payload);
+    String budibaseUserId = (String) jsonObject.get("userId");
+    return budibaseUserId.substring(3);
+  }
+
+  public void validateConsultantRequest(String body, HttpMethod method,
+      HttpServletRequest request) {
+    String consultantId = extractUserIdFromJWT(request);
+
+    String userId;
+    if (request.getRequestURI().contains("api/v2/queries")) {
+      userId = extractUserIdFromV2Query(body);
+    } else {
+      userId = extractUserIdFromBodyReadOperation(method, body);
+    }
+
+    if (consultantId.equals(userId)) {
+      return;
+    }
+
+    List<String> consultantAssignedUsers = budibaseApiService
+        .getConsultantAssignedUsers(consultantId);
+    Optional<String> matchedUserId = consultantAssignedUsers.stream()
+        .filter(el -> el.equals(userId)).findFirst();
+    if (matchedUserId.isEmpty()) {
+      throw new NotAllowedException(
+          "You don't have permissions to access user specific data for user with id: " + userId
+      );
+    }
+  }
+
+
+  @Cacheable(cacheNames = CacheManagerConfig.TOKEN_CACHE)
+  public String extractRolesOfCurrentUsers(String queryStirng, String cookie) {
+    URI uri = null;
+    try {
+      uri = new URI("http", null, proxyServiceHost, proxyServicePort, null, null, null);
+      uri = UriComponentsBuilder.fromUri(uri)
+          .path("api/global/self")
+          .query(queryStirng)
+          .build(true).toUri();
+    } catch (URISyntaxException e) {
+      //
```

Comment:
unhandled exception, please at least log it

Reply:
done